src/DataProvider/OperationDataProviderTrait.php

```diff
@@ -0,0 +1,128 @@
+<?php
+
+/*
+ * This file is part of the API Platform project.
+ *
+ * (c) Kévin Dunglas <dunglas@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace ApiPlatform\Core\DataProvider;
+
+use ApiPlatform\Core\Exception\InvalidIdentifierException;
+use ApiPlatform\Core\Identifier\Normalizer\ChainIdentifierDenormalizer;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpKernel\Exception\NotFoundHttpException;
+
+trait OperationDataProviderTrait
+{
+    /**
+     * @var CollectionDataProviderInterface
+     */
+    private $collectionDataProvider;
+    /**
+     * @var ItemDataProviderInterface
+     */
+    private $itemDataProvider;
+    /**
+     * @var SubresourceDataProviderInterface
+     */
+    private $subresourceDataProvider;
+    /**
+     * @var ChainIdentifierDenormalizer
+     */
+    private $identifierDenormalizer;
+
+    /**
+     * Retrieves data for a collection operation.
+     *
+     * @return array|\Traversable|null
+     */
+    private function getCollectionData(Request $request, array $attributes, array $context)
```

Comment:
Don't you think we can avoid to depend of the request, but extract the relevant data from the controller/event listener and pass it through the context? It will be easier to support PSR-7 this way.

Reply:
Yes we can totally do this will be cleaner.